### PR TITLE
Later merged

### PR DESCRIPTION
**Pull request by Saizan**

### CODE REVIEW

File: src/full/Agda/TypeChecking/Free/Old.hs [hunk 1]
```diff
@@ -0,0 +1,358 @@
+
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Free/Old.hs]: Did you bring back this old file by intention?

**Review by andreasabel — CHANGES_REQUESTED** [review]
Implementing guarded type theory?

**Comment by Saizan** [comment]
> Implementing guarded type theory?

Yes! I actually had this done a while ago.

File: src/full/Agda/TypeChecking/Free/Old.hs (hunk 1 above)
Comment (Saizan) [reply src/full/Agda/TypeChecking/Free/Old.hs]: No, I guess it happened during the merge.

**Comment by andreasabel** [comment]
I guess I wait with a review until this PR is ready and a few examples have been added to the testsuite.

**Comment by Saizan** [comment]
Examples: https://github.com/agda/agda/blob/0314824df5dddcfe80b384c23763263a5ef59e8b/test/Succeed/LaterPrims.agda#L77

**Comment by Saizan** [comment]
@andreasabel do you think you'll have time to review soon? Otherwise it might make sense to merge before too many conflicts accumulate

**Comment by andreasabel** [comment]
> @andreasabel do you think you'll have time to review soon? Otherwise it might make sense to merge before too many conflicts accumulate

Er, I agree, please merge this, I will probably not be able to say anything meaningful soon.

File: src/full/Agda/TypeChecking/Lock.hs-boot [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+checkLockedVars
+  :: Term
+     -- ^ term to check
+  -> Type
+     -- ^ its type
+  -> Arg Term
+     -- ^ the lock
+  -> Type
+     -- ^ type of the lock
+  -> TCM ()
```
Comment (ice1000) [inline src/full/Agda/TypeChecking/Lock.hs-boot:36]: I don't think we need this file, do we?

Comment (Saizan) [reply src/full/Agda/TypeChecking/Lock.hs-boot:36]: Are you saying `Agda.TypeChecking.Constraints` can just `import Agda.TypeChecking.Lock` without an import cycle?

Comment (ice1000) [reply src/full/Agda/TypeChecking/Lock.hs-boot:36]: Oh, Ok...

Comment (ice1000) [reply src/full/Agda/TypeChecking/Lock.hs-boot:36]: I missed that

**Comment by L-TChen** [comment]
Is this based on the paper Guarded Cubical Type Theory (https://link.springer.com/10.1007/s10817-018-9471-7)?

**Comment by Saizan** [comment]
More closely based on https://dl.acm.org/doi/abs/10.1145/3372885.3373814

We use tick variables rather than delayed substitutions.

**Comment by L-TChen** [comment]
Right, thanks! 

As for the pull request itself, shouldn't there be at least a list of relevant references in the official documentation?

**Comment by Saizan** [comment]
Yeah, there should be documentation in general.

**Pull request merged**